### PR TITLE
Client close function

### PR DESCRIPTION
**Pull request by majordoobie**
#111 
[Added coc.Client.close_client coroutine] Added coroutine to allow us…ers to close the client connection from within a async function such as async main(). This is useful to allow users to use the `asyncio.run()` wrapper properly

### CODE REVIEW

File: coc/client.py
Comment (r-priyam) [inline coc/client.py:292]:
```suggestion
        LOG.info("Clash of Clans client logging out...")
        self.dispatch("on_client_close")
        await self.http.close()
```
`on_client_close` event should dispatch here instead of in `close_loops`

File: coc/client.py
Comment (r-priyam) [inline coc/client.py:287]:
```suggestion
    def close_loops(self) -> None:
        """Closes the HTTP connection
        """
        self.loop.run_until_complete(self.http.close())
        self.loop.close()
```

**Review by r-priyam — CHANGES_REQUESTED** [review]

File: coc/client.py
Comment (r-priyam) [reply coc/client.py:287]: > Why are you omitting the dispatch?

because this is actually closing the `asyncio loop` and not doing what the dispatch name says. The thing that dispatch name says is actually being one by the `close_client` method

File: README.rst
Comment (mathsman5133) [inline README.rst:75]: This line should probably be indented

File: coc/client.py
Comment (mathsman5133) [inline coc/client.py:281]: Only little thing I'm worried about is making sure we don't break stuff in a non-major version, ie. I think (please correct me if I'm wrong) but client.close() now doesn't exist. 

Long-term, should we always be creating and closing clients with await stuff and letting asyncio handle running async stuff in non-async places? I wonder whether that would solve all our issues, and whether I'm trying to do something that isn't really my job.

So change `.close()` to be an async function and either make `coc.login()` an async function or remove it completely.

Like you showed, it's easier (and better if it doesn't have all those loop cleanup issues) to leave it to asyncio.run() to run it all. Thoughts? I'm kinda just rambling. Either way, I think we can add this stuff with different names now, but need to keep the existing functions until v3 (maybe adding a depreciated warning)

**Review by mathsman5133 — COMMENTED** [review]
Thanks. I think you've found the proper way to fix all these loop cleanup issues :)

File: coc/client.py
Comment (doluk) [reply coc/client.py:281]: asyncio.run() requieres python 3.7 or newer, we would need to add that somewhere too.

The big downside with asyncio.run() is most likely the usage of a coc client with another aiohttp session, like a discord bot. We should definitely test that before we discussing this solution further.

Comment (mathsman5133) [reply coc/client.py:281]: Absolutely, and very true re. python 3.7

I guess the `asyncio.run` stuff would only be for people running it as a standalone script thing - else the current `coc.login` is fine (and doesn't have the loop cleanup issues / people don't care as much because there's a hundred other warnings when you close a bot). I think the real issue is `client.close()`, not login so much. I don't think people generally have ever had many issues with login in it's current form

**Comment by doluk** [comment]
the example code with the modifications above produces still an error.
```
Exception ignored in: <function _ProactorBasePipeTransport.__del__ at 0x0000018108223820>
Traceback (most recent call last):
  File "C:\Users\doluk\AppData\Local\Programs\Python\Python39\lib\asyncio\proactor_events.py", line 116, in __del__
    self.close()
  File "C:\Users\doluk\AppData\Local\Programs\Python\Python39\lib\asyncio\proactor_events.py", line 108, in close
    self._loop.call_soon(self._call_connection_lost, None)
  File "C:\Users\doluk\AppData\Local\Programs\Python\Python39\lib\asyncio\base_events.py", line 746, in call_soon
    self._check_closed()
  File "C:\Users\doluk\AppData\Local\Programs\Python\Python39\lib\asyncio\base_events.py", line 510, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
```
According to [this aiohttp issue](https://github.com/aio-libs/aiohttp/issues/1925), [this aiohttp issue](https://github.com/aio-libs/aiohttp/issues/6635) and [this httpx issue](https://github.com/encode/httpx/issues/914) the problem is the underlaying SSL transport. Currently the only way to avoid this error message is to give the SSL transport time to shutdown by adding a `time.sleep(0.25)` or `await asyncio.sleep(0.25)` in coc.http.HTTPClient.close(). With that, no exception gets raised executing the sample code.
Another fix for that would be switch from aiohttp to trio as mentioned in the issues, but yeah...

The example bot is also affected by the same error but not as you may think. The discord bot itself has the same problem when using `await bot.close()`. Here also the "same" workaround `time.sleep(0.25)` solves the problem.
The added line in coc.http.HTTPClient.close() does not affect the bot (and to be honest most of us don't even logout the coc client). So we should be able to merge this fix with the described fixed without causing further issues.

**Comment by doluk** [comment]
as soon as @majordoobie merges the [pr](https://github.com/majordoobie/coc.py/pull/1) we should be good to merge this one too

File: coc/client.py
Comment (majordoobie) [reply coc/client.py:281]: So you actually mentioned it I personally do not think it is on your library to handle the cleanup because it will create a race condition with other libraries that also depend on sessions to be cleaned out. 

This is the best way I have found to handle cleanup of both coc.py sessions and asyncpg and that's to drop into "manual" loops instead of the abstracted run only to be able to properly clean up after myself. The asyncio.run does do the generator cleanup but it will not properly clean up sessions. Which may or may not be an issue. 

https://github.com/majordoobie/PantherLily/blob/main/main.py#L117

**Comment by majordoobie** [comment]
Moved to #116

**Pull request closed without merging**
